cub/cub/thread/thread_scan.cuh

```diff
@@ -164,6 +164,86 @@ ThreadScanExclusive(T (&input)[LENGTH], T (&output)[LENGTH], ScanOp scan_op, T p
   return ThreadScanExclusive<LENGTH>((T*) input, (T*) output, scan_op, prefix, apply_prefix);
 }
 
+/**
+ * @brief Perform a sequential exclusive prefix scan over the first @p valid_items elements
+ *        of the statically-sized @p input array, seeded with the specified @p prefix.
+ *
+ * @tparam Input
+ *   <b>[inferred]</b> The data type holding the input data having member
+ *   <tt>operator[](int i)</tt> and must be statically-sized (<tt>size()</tt> method or static array)
+ *
+ * @tparam Output
+ *   <b>[inferred]</b> The data type to hold the scan result having member
+ *   <tt>operator[](int i)</tt> and must be statically-sized (<tt>size()</tt> method or static array)
+ *
+ * @tparam ScanOp
+ *   <b>[inferred]</b> Binary scan operator type having member
+ *   <tt>T operator()(const T &a, const T &b)</tt>
+ *
+ * @tparam PrefixT
+ *   <b>[inferred]</b> The data type of the prefix.
+ *
+ * @param[in] input
+ *   Array-like input
+ *
+ * @param[out] output
+ *   Array-like output (may be aliased to @p input)
+ *
+ * @param[in] scan_op
+ *   Binary scan operator
+ *
+ * @param[in] valid_items
+ *   Number of valid items
+ *
+ * @param[in] prefix
+ *   Prefix to seed scan with
+ *
+ * @param[in] apply_prefix
+ *   Whether or not the calling thread should apply its prefix.
+ *   (Handy for preventing thread-0 from applying a prefix.)
+ */
+template <typename Input,
+          typename Output,
+          typename ScanOp,
+          typename PrefixT,
+          typename ValueT = _CUDA_VSTD::iter_value_t<Input>,
+          typename AccumT = _CUDA_VSTD::__accumulator_t<ScanOp, ValueT, PrefixT>>
+_CCCL_DEVICE _CCCL_FORCEINLINE void ThreadScanExclusivePartial(
+  Input& input, Output& output, ScanOp scan_op, int valid_items, PrefixT prefix, bool apply_prefix = true)
+{
+  using namespace cub::detail;
+  static_assert(is_fixed_size_random_access_range_v<Input>,
+                "Input must support the subscript operator[] and have a compile-time size");
+  static_assert(is_fixed_size_random_access_range_v<Output>,
+                "Output must support the subscript operator[] and have a compile-time size");
+  static_assert(has_binary_call_operator<ScanOp, ValueT>::value,
+                "ScanOp must have the binary call operator: operator(ValueT, ValueT)");
+  constexpr auto length = static_size_v<Input>;
+  static_assert(static_size_v<Output> == length);
+  AccumT inclusive = input[0];
+  if (valid_items > 0)
+  {
+    if (apply_prefix)
+    {
+      inclusive = scan_op(prefix, inclusive);
+    }
+    output[0] = prefix;
+  }
+
+  AccumT exclusive = inclusive;
+
+  _CCCL_PRAGMA_UNROLL_FULL()
+  for (int i = 1; i < length; ++i)
+  {
+    if (i < valid_items)
+    {
+      inclusive = scan_op(exclusive, input[i]);
```

Comment:
important: the current implementation of the exclusive scan appears to invoke `scan_op` one additional time compared to what's strictly necessary, as the `inclusive` result is ultimately discarded.

Is addressing this redundant invocation something that would significantly broaden the scope of your current work, or do you anticipate it being limited to code paths that will already be modified for the `valid_items` implementation?
```suggestion
      inclusive = scan_op(exclusive, input[i]);
```

Reply:
Also, if we decide to address this, we should add targeted tests that would fail with the current behavior to ensure the fix is effective and prevents regressions.

Reply:
Given that we later decided to guarantee nothing about the output of `ThreadReducePartial` with `valid_items < 1`, I would rather be in favor of changing `ThreadScan*Partial` back to actually return the aggregate for consistency. Then it is also clear from the API that the operator will also be applied to the last valid element in even in an exclusive scan. Although, there the question remains if `valid_items < 1` with `apply_prefix == true` should mean that we have to return the prefix as the aggregate or if it is ok to return anything in that case which would mean no code changes.